qiime2/sdk/proxy.py

```diff
@@ -0,0 +1,193 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2023, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+from qiime2.core.type.util import is_visualization_type, is_collection_type
+
+
+class Proxy:
+    """Abstract base class to indicate that a given class that inherits from it
+    is a proxy
+    """
+    pass
+
+
+class ProxyResult(Proxy):
+    def __init__(self, future, selector, signature=None):
+        """We have a future that represents the results of some QIIME 2 action,
+        and we have a selector indicating specifically which result we want
+        """
+        self._future_ = future
+        self._selector_ = selector
+        self._signature_ = signature
+
+    @property
+    def _archiver(self):
+        return self.result()._archiver
+
+    @property
+    def type(self):
+        return self._archiver.type
```

Comment:
Strictly speaking, this doesn't need to block, as the type can be resolved from `signature.solve_outputs` + selector. And I _believe_ enough information exists to make that happen before anything gets scheduled by the parsl app.

Reply:
You can tell I just copy-pasted this from Result because otherwise I would have at least used the approach from the repr lol